google/cloud/spanner/oid.h

```diff
@@ -0,0 +1,81 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_OID_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_OID_H
+
+#include "google/cloud/spanner/version.h"
+#include <ostream>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A representation of the PostgreSQL Object Identifier type: an opaque string
+ * used as primary keys for various system tables in the PostgreSQL dialect.
+ */
+class PgOid {
+ public:
+  /// Default construction yields OID 0.
+  PgOid() : rep_("0") {}
+
+  /// @name Regular value type, supporting copy, assign, move.
+  ///@{
+  PgOid(PgOid const&) = default;
+  PgOid& operator=(PgOid const&) = default;
+  PgOid(PgOid&&) = default;
+  PgOid& operator=(PgOid&&) = default;
+  ///@}
+
+  /**
+   * Construction from an oid string. Note that there is no check here that the
+   * argument string is indeed well-formatted. Error detection will be delayed
+   * until the value is passed to Spanner.
+   */
+  explicit PgOid(std::string value) : rep_(std::move(value)) {}
+
+  /// @name Conversion to a oid string.
+  ///@{
+  explicit operator std::string() const& { return rep_; }
+  explicit operator std::string() && { return std::move(rep_); }
+  ///@}
+
+ private:
+  std::string rep_;
```

Comment:
https://www.postgresql.org/docs/current/datatype-oid.html#DATATYPE-OID

> The `oid` type is currently implemented as an unsigned four-byte integer.

String may be the right choice, but maybe deserves a comment?

Reply:
A normal integer is a string on the wire too, but we don't make the integral C++ type a string.  So if uint64 is the natural representation of an OID, then perhaps that should be the `rep_` type.  Why have construction from and conversion to a string at all?

Reply:
Changed the representation to uint64_t, but we still need the conversion to a string in order to serialize the value

Reply:
Yes, we do need to convert to string at some point, but it can be deferred until serialization time.  So, `PgOid` would just have...
```
  explicit operator std::uint64_t() const { return rep_; }
```
Then, comparisons would be directly by value, without having to convert to a string representation first.

And the output operator would be just...
```
  return os << std::uint64_t(oid);
```